snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,309 @@
+from threading import Event
+from typing import (
+    Callable,
+    Generic,
+    Iterable,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Set,
+)
+
+from snuba.utils.concurrent import Synchronized, execute
+from snuba.utils.streams.consumers.backends.abstract import ConsumerBackend
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+    def get_effective_remote_consumer_group_offset(
+        self, stream: TStream, groups: Iterable[ConsumerGroup]
+    ) -> Optional[TOffset]:
+        offsets = self.remote_consumer_group_offsets[stream]
+        return min(
+            filter(
+                lambda offset: offset is not None,
+                [offsets.get(group) for group in groups],
+            ),
+            default=None,
+        )
+
+
+class SynchronizedConsumerBackend(ConsumerBackend[TStream, TOffset, TValue]):
+    def __init__(
+        self,
+        backend: ConsumerBackend[TStream, TOffset, TValue],
+        commit_log_consumer_backend: ConsumerBackend[
+            # NOTE: These TStream and TOffset values for the commit log
+            # consumer do not need to match, but this simplifies things a bit
+            # so that we don't have to declare two additional type variables.
+            TStream,
+            TOffset,
+            CommitData[TStream, TOffset],
+        ],
+        remote_consumer_groups: Set[ConsumerGroup],
+        get_commit_log_streams_for_topic: Callable[[Topic[TStream]], Sequence[TStream]],
+    ) -> None:
+        self.__backend = backend
+        self.__commit_log_consumer_backend = commit_log_consumer_backend
+        self.__remote_consumer_groups = remote_consumer_groups
+        self.__get_commit_log_streams_for_topic = get_commit_log_streams_for_topic
+
+        self.__subscription: Synchronized[
+            Subscription[TStream, TOffset]
+        ] = Synchronized(Subscription([]))
+
+        self.__paused_streams: Set[TStream] = set()
+
+        self.__shutdown_requested = Event()
+
+        # TODO: If the commit log consumer crashes, it should cause all method
+        # calls on this object to raise (probably a RuntimeError raised from
+        # the commit log consumer exception.)
+        self.__commit_log_consumer_future = execute(self.__run_commit_log_consumer)
```

Comment:
Do you ever need to subscribe on the commit_log_consumer? I don;t see any call to subscribe

Reply:
The commit log consumer doesn't subscribe as part of a consumer group (balanced asignment) — it is manually assigned the streams to consume from (managed assignment), so instead of `subscribe` it uses [`assign` on line 117](https://github.com/getsentry/snuba/pull/600/files#diff-5f2067cb2705eb83733b6b81d2eeb2faR117) (this maps to [the Kafka driver's `assign` method](https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.Consumer.assign).)

I'm not fixated on that idea though, [per this comment]( https://github.com/getsentry/snuba/pull/600/files#r351067888). 

This decision one way or another doesn't really reduce the amount of work that needs to happen though, ultimately the subscription consumer will need to use managed assignment, at least in the long-term.